astropy/nddata/decorators.py

```diff
@@ -152,10 +152,10 @@ def support_nddata_decorator(func):
             if param.kind in (param.VAR_POSITIONAL, param.VAR_KEYWORD):
                 raise ValueError("func may not have *args or **kwargs.")
             try:
-                if param.default == param.empty:
-                    func_args.append(param_name)
-                else:
+                if param.kind == param.KEYWORD_ONLY or param.default != param.empty:
```

Comment:
This seems incorrect: a positional-only argument with a default value would end up in `func_kwargs` and would produce the same kind of bug that you're fixing here (in fact, given the old logic, this bug already exists). I think we should add a test case for it and fix it here too.

Reply:
Oh yes thanks. Forgot all about positional only.

Reply:
oh god @neutrinoceros you have unleashed a kracken with this one. There are so many edge cases when you get into positional arguments.

Reply:
I think I got it.

Reply:
sorry about that !
I'll review again now 🐙